HeterogeneousCore/SonicTriton/interface/triton_utils.h

```diff
@@ -11,16 +12,73 @@
 #include "grpc_client.h"
 
 namespace triton_utils {
-
-  using Error = triton::client::Error;
-
   template <typename C>
   std::string printColl(const C& coll, const std::string& delim = ", ");
-
-  //helper to turn triton error into exception
-  void throwIfError(const Error& err, std::string_view msg);
+  //implemented as a standalone function to avoid repeated specializations for different TritonData types
+  template <typename DT>
+  bool checkType(inference::DataType dtype) {
+    return false;
+  }
 }  // namespace triton_utils
 
+//explicit specializations (inlined)
+//special cases:
+//bool: vector<bool> doesn't have data() accessor, so use char (same byte size)
+//FP16 (half precision): no C++ primitive exists, so use uint16_t (e.g. with libminifloat)
+template <>
+inline bool triton_utils::checkType<char>(inference::DataType dtype) {
+  return dtype == inference::DataType::TYPE_BOOL or dtype == inference::DataType::TYPE_STRING;
+}
+template <>
+inline bool triton_utils::checkType<uint8_t>(inference::DataType dtype) {
+  return dtype == inference::DataType::TYPE_UINT8;
+}
+template <>
+inline bool triton_utils::checkType<uint16_t>(inference::DataType dtype) {
+  return dtype == inference::DataType::TYPE_UINT16 or dtype == inference::DataType::TYPE_FP16;
+}
+template <>
+inline bool triton_utils::checkType<uint32_t>(inference::DataType dtype) {
+  return dtype == inference::DataType::TYPE_UINT32;
+}
+template <>
+inline bool triton_utils::checkType<uint64_t>(inference::DataType dtype) {
+  return dtype == inference::DataType::TYPE_UINT64;
+}
+template <>
+inline bool triton_utils::checkType<int8_t>(inference::DataType dtype) {
+  return dtype == inference::DataType::TYPE_INT8;
+}
+template <>
+inline bool triton_utils::checkType<int16_t>(inference::DataType dtype) {
+  return dtype == inference::DataType::TYPE_INT16;
+}
+template <>
+inline bool triton_utils::checkType<int32_t>(inference::DataType dtype) {
+  return dtype == inference::DataType::TYPE_INT32;
+}
+template <>
+inline bool triton_utils::checkType<int64_t>(inference::DataType dtype) {
+  return dtype == inference::DataType::TYPE_INT64;
+}
+template <>
+inline bool triton_utils::checkType<float>(inference::DataType dtype) {
+  return dtype == inference::DataType::TYPE_FP32;
+}
+template <>
+inline bool triton_utils::checkType<double>(inference::DataType dtype) {
+  return dtype == inference::DataType::TYPE_FP64;
+}
+
+//helper to turn triton error into exception
+//implemented as a macro to avoid constructing the MSG string for successful function calls
+#define tritonThrowIfError(X, MSG)                                                                            \
```

Comment:
An alternative could be to have a templated function that takes a function to craft the message. With lambdas and inlining I'd expect the compiler to emit code similar to here. You could also have the function to take `TritonException` argument, so it could stream the message parts directly there without constructing a temporary string first.

Reply:
Do you have a preference for one of these vs. the current approach? While I agree with the general principle of avoiding complicated macros, this one is really quite simple, and seems like it might be easier to read/maintain than a lambda-based approach. (I'm not sure streaming would help: one of the contributors was the `to_string()` call in `TritonInputHeapResource::copyInput()`, and it seems like a similar conversion would have to happen if streaming was used.)

Reply:
I don't have a strong preference (especially since this is an internal function/macro not exposed to users). I would suggest to follow the all-caps naming convention, because the macro expands to something else than a function call, and therefore can't be used in all context's where a function could be.

Reply:
Done.